### PR TITLE
Adjust to consul Api changes

### PR DESCRIPTION
**Pull request by malfoj89**
Consul api has changed and some fields are not returned when not set which causes to an exception being thrown however the behavior of the api is correct. Making those fields as optionals

### CODE REVIEW

File: src/main/java/pl/allegro/tech/discovery/consul/recipes/watch/catalog/ServiceInstance.java [hunk 1]
```diff
@@ -1,17 +1,19 @@
...
     private final List<String> serviceTags;
     private final String serviceAddress;
-    private final int servicePort;
+    private final Integer servicePort;
 
-    public ServiceInstance(String serviceId, List<String> serviceTags, String serviceAddress, int servicePort) {
+    public ServiceInstance(String serviceId, List<String> serviceTags, String serviceAddress, Integer servicePort) {
         this.serviceId = serviceId;
-        this.serviceTags = serviceTags;
+        this.serviceTags = serviceTags != null ? serviceTags : Collections.emptyList();
```
Comment (Ferdudas97) [inline src/main/java/pl/allegro/tech/discovery/consul/recipes/watch/catalog/ServiceInstance.java:16]: maybe it also should be nullable?

**Review by Ferdudas97 — DISMISSED** [review]

File: src/main/java/pl/allegro/tech/discovery/consul/recipes/json/JsonValueReader.java [hunk 2]
```diff
@@ -11,10 +11,25 @@ public static <T> T requiredValue(Map serviceProps, String property, Class<T> cl
             throw new JsonDecoder.JsonDecodeException(property + " property is missing in JSON. " +
                     "This may indicate that there are incompatible changes in Consul API.");
         }
+        mustBeAbleToCastToClass(property, clazz, value);
+        return (T) value;
+    }
+
+
+    @SuppressWarnings("unchecked")
+    public static <T> T optionalValue(Map serviceProps, String property, Class<T> clazz) {
```
Comment (matb4r) [inline src/main/java/pl/allegro/tech/discovery/consul/recipes/json/JsonValueReader.java:20]: I guess this method should be named more like `nullableValue`, or return Optional (and do not build it in `ServiceInstance` then)

File: src/main/java/pl/allegro/tech/discovery/consul/recipes/watch/catalog/ServiceInstance.java [hunk 3]
```diff
@@ -24,12 +26,12 @@ public List<String> getServiceTags() {
         return serviceTags;
     }
 
-    public String getServiceAddress() {
-        return serviceAddress;
+    public Optional<String> getServiceAddress() {
+        return Optional.ofNullable(serviceAddress);
```
Comment (matb4r) [inline src/main/java/pl/allegro/tech/discovery/consul/recipes/watch/catalog/ServiceInstance.java:30]: I guess that hashCode/equals do not need to be updated to use `.get()`?

File: src/main/java/pl/allegro/tech/discovery/consul/recipes/watch/catalog/ServiceInstance.java [hunk 4]
```diff
@@ -24,12 +26,12 @@ public List<String> getServiceTags() {
         return serviceTags;
     }
 
-    public String getServiceAddress() {
-        return serviceAddress;
+    public Optional<String> getServiceAddress() {
```
Comment (matb4r) [inline src/main/java/pl/allegro/tech/discovery/consul/recipes/watch/catalog/ServiceInstance.java:29]: Where are we using this getters?

File: src/test/groovy/pl/allegro/tech/discovery/consul/recipes/watch/health/HealthServiceInstancesWatcherIntTest.groovy [hunk 5]
```diff
@@ -50,19 +57,46 @@ class HealthServiceInstancesWatcherIntTest extends Specification {
...
+        when: "unhealthy service instance is registered"
+        consulCluster.registerInstanceLackingPortNumber("my-service", GIVEN_DC, GIVEN_NODE, ["tag1", "tag2"])
+
+        then: "watcher caught new state with only healthy service instance"
+        new PollingConditions(timeout: 10).eventually {
+            latestState.head().instances.size() == 1
+            def instanceWithNoPort = latestState.head().instances.first()
+            instanceWithNoPort.serviceId != null
+            instanceWithNoPort.serviceAddress.get() == "localhost"
+            !instanceWithNoPort.servicePort.isPresent()
```
Comment (matb4r) [inline src/test/groovy/pl/allegro/tech/discovery/consul/recipes/watch/health/HealthServiceInstancesWatcherIntTest.groovy:98]:
```suggestion
            instanceWithNoPort.servicePort.isEmpty()
```

File: src/test/groovy/pl/allegro/tech/discovery/consul/recipes/watch/health/HealthServiceInstancesWatcherIntTest.groovy [hunk 6]
```diff
@@ -50,19 +57,46 @@ class HealthServiceInstancesWatcherIntTest extends Specification {
...
 
         then: "watcher caught new state with only healthy service instance"
         new PollingConditions(timeout: 10).eventually {
             latestState.head().instances.size() == 1
             def instance = latestState.head().instances.first()
             instance.serviceId != null
-            instance.serviceAddress == "localhost"
-            instance.servicePort == 1234
+            instance.serviceAddress.get() == "localhost"
+            instance.servicePort.get() == 1234
```
Comment (KSmigielski) [inline src/test/groovy/pl/allegro/tech/discovery/consul/recipes/watch/health/HealthServiceInstancesWatcherIntTest.groovy:71]: I don't like using `Optional#get` without checking if value is presnt. This way when something work wrong in tests we will get NPE. I suggest to compare it to `Optional` or checking presents before getting value.
Similar with other assertion with `Optional`

**Review by KSmigielski — CHANGES_REQUESTED** [review]

File: src/main/java/pl/allegro/tech/discovery/consul/recipes/watch/catalog/ServiceInstance.java (hunk 3 above)
Comment (malfoj89) [reply src/main/java/pl/allegro/tech/discovery/consul/recipes/watch/catalog/ServiceInstance.java:30]: indeed not need

File: src/main/java/pl/allegro/tech/discovery/consul/recipes/watch/catalog/ServiceInstance.java (hunk 4 above)
Comment (malfoj89) [reply src/main/java/pl/allegro/tech/discovery/consul/recipes/watch/catalog/ServiceInstance.java:29]: in the implementation of this lib

File: src/test/groovy/pl/allegro/tech/discovery/consul/recipes/watch/health/HealthServiceInstancesWatcherIntTest.groovy (hunk 5 above)
Comment (malfoj89) [reply src/test/groovy/pl/allegro/tech/discovery/consul/recipes/watch/health/HealthServiceInstancesWatcherIntTest.groovy:98]: groovy has issue with that, and thats why it has been done this way `No signature of method: java.util.Optional.isEmpty() is applicable for argument types: () values: []
Possible solutions: empty(), identity(groovy.lang.Closure), inspect(), dump()
groovy.lang.MissingMethodException: No signature of method: java.util.Optional.isEmpty() is applicable for argument types: () values: []
Possible solutions: empty(), identity(groovy.lang.Closure), inspect(), dump()
	at pl.allegro.tech.discovery.consul.recipes.watch.health.HealthServiceInstancesWatcherIntTest.should watch only parsable instances(HealthServiceInstancesWatcherIntTest.groovy:93)`

File: src/test/groovy/pl/allegro/tech/discovery/consul/recipes/watch/health/HealthServiceInstancesWatcherIntTest.groovy (hunk 6 above)
Comment (malfoj89) [reply src/test/groovy/pl/allegro/tech/discovery/consul/recipes/watch/health/HealthServiceInstancesWatcherIntTest.groovy]: changed check now

File: src/main/java/pl/allegro/tech/discovery/consul/recipes/json/JsonValueReader.java (hunk 2 above)
Comment (malfoj89) [reply src/main/java/pl/allegro/tech/discovery/consul/recipes/json/JsonValueReader.java]: renamed as proposed

**Review by matb4r — APPROVED** [review]

**Review by KSmigielski — APPROVED** [review]

**Pull request merged**